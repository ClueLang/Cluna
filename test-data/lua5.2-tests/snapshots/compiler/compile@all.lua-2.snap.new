---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
