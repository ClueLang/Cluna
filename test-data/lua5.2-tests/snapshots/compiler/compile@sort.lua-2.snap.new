---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/sort.lua
---
print("testing (parts of) table library");
print("testing unpack");
local unpack = table.unpack
local x, y, z, a, n
a = {}
lim = 2000
for i = 1, lim {
    a[(i)] = i
}
assert(select(lim, unpack(a)) == lim && select('#', unpack(a)) == lim);
x = unpack(a)
assert(x == 1);
x = {
    unpack(a)
}
assert(#x == lim && x[(1)] == 1 && x[(lim)] == lim);
x = {
    unpack(a, lim - 2)
}
assert(#x == 3 && x[(1)] == lim - 2 && x[(3)] == lim);
x = {
    unpack(a, 10, 6)
}
assert(next(x) == nil);
x = {
    unpack(a, 11, 10)
}
assert(next(x) == nil);
x, y = unpack(a, 10, 10)
assert(x == 10 && y == nil);
x, y, z = unpack(a, 10, 11)
assert(x == 10 && y == 11 && z == nil);
a, x = unpack({
    1
})
assert(a == 1 && x == nil);
a, x = unpack({
    1, 
    2
}, 1, 1)
assert(a == 1 && x == nil);
if !_no32 {
    assert(!pcall(unpack, {}, 0, 2 ^ 31 - 1));
    assert(!pcall(unpack, {}, 1, 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31), 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31 - 1), 2 ^ 31 - 1));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 0));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 1));
    pcall(unpack, {}, 1, 2 ^ 31);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 1, 2 ^ 31 - 1)
    assert(a == 20 && b == nil);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 2, 2 ^ 31 - 1)
    assert(a == nil && b == 20);
}
print("testing pack");
a = table.pack()
assert(a[(1)] == nil && a.n == 0);
a = table.pack(table)
assert(a[(1)] == table && a.n == 1);
a = table.pack(nil, nil, nil, nil)
assert(a[(1)] == nil && a.n == 4);
print("testing sort");
local fn check(t) {
    local fn f(a, b) {
        assert(a && b);
        return true
    }
    local s, e = pcall(table.sort, t, f)
    assert(!s && e::find("invalid order function"));
}
check({
    1, 
    2, 
    3, 
    4
});
check({
    1, 
    2, 
    3, 
    4, 
    5
});
check({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
global fn check(a, f) {
    f = f || fn (x, y) {
        return x < y
    }
    for n = #a, 2,  - 1 {
        assert(!f(a[(n)], a[(n - 1)]));
    }
}
a = {
    "Jan", 
    "Feb", 
    "Mar", 
    "Apr", 
    "May", 
    "Jun", 
    "Jul", 
    "Aug", 
    "Sep", 
    "Oct", 
    "Nov", 
    "Dec"
}
table.sort(a);
check(a);
global fn perm(s, n) {
    n = n || #s
    if n == 1 {
        local t = {
            unpack(s)
        }
        table.sort(t);
        check(t);
    } else {
        for i = 1, n {
            s[(i)], s[(n)] = s[(n)], s[(i)]
            perm(s, n - 1);
            s[(i)], s[(n)] = s[(n)], s[(i)]
        }
    }
}
perm({});
perm({
    1
});
perm({
    1, 
    2
});
perm({
    1, 
    2, 
    3
});
perm({
    1, 
    2, 
    3, 
    4
});
perm({
    2, 
    2, 
    3, 
    4
});
perm({
    1, 
    2, 
    3, 
    4, 
    5
});
perm({
    1, 
    2, 
    3, 
    3, 
    5
});
perm({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
perm({
    2, 
    2, 
    3, 
    3, 
    5, 
    6
});
limit = 30000
if _soft {
    limit = 5000
}
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
local x = os.clock()
table.sort(a);
print(string.format("Sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
x = os.clock()
table.sort(a);
print(string.format("Re-sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
x = os.clock()
i = 0
table.sort(a, fn (x, y) {
    i = i + 1
    return y < x
});
print(string.format("Invert-sorting other %d elements in %.2f sec., with %i comparisons", limit, os.clock() - x, i));
check(a, fn (x, y) {
    return y < x
});
table.sort({});
for i = 1, limit {
    a[(i)] = false
}
x = os.clock()
table.sort(a, fn (x, y) {
    return nil
});
print(string.format("Sorting %d equal elements in %.2f sec.", limit, os.clock() - x));
check(a, fn (x, y) {
    return nil
});
for i, v with pairs(a) {
    assert(!v || i == 'n' && v == limit);
}
A = {
    "Ã¡lo", 
    "\0first :-)", 
    "alo", 
    "then this one", 
    "45", 
    "and a new"
}
table.sort(A);
check(A);
table.sort(A, fn (x, y) {
    load(string.format("A[%q] = ''", x))();
    collectgarbage();
    return x < y
});
tt = {
    __lt = fn (a, b) {
        return a.val < b.val
    }
}
a = {}
for i = 1, 10 {
    a[(i)] = {
        val = math.random(100)
    }
    setmetatable(a[(i)], tt);
}
table.sort(a);
check(a, tt.__lt);
check(a);
print("OK");
