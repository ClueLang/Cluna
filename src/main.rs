#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
