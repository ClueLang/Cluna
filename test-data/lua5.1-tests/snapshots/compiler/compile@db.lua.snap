---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.1-tests/db.lua
---
local fn dostring(s) {
    return assert(loadstring(s))()
}
print("testing debug library and debug information");
{
    local a = 1
}
global fn test(s, l, p) {
    collectgarbage();
    local fn f(event, line) {
        assert(event == 'line');
        local l = table.remove(l, 1)
        if p {
            print(l, line);
        }
        assert(l == line, "wrong trace!!");
    }
    debug.sethook(f, "l");
    loadstring(s)();
    debug.sethook();
    assert(table.getn(l) == 0);
}
{
    local a = debug.getinfo(print)
    assert(a.what == "C" && a.short_src == "[C]");
    local b = debug.getinfo(test, "SfL")
    assert(b.name == nil && b.what == "Lua" && b.linedefined == 11 && b.lastlinedefined == b.linedefined + 10 && b.func == test && notstring.find(b.short_src, "%["));
    assert(b.activelines[(b.linedefined + 1)] && b.activelines[(b.lastlinedefined)]);
    assert(notb.activelines[(b.linedefined)] && notb.activelines[(b.lastlinedefined + 1)]);
}
a = "function f () end"
local fn dostring(s, x) {
    return loadstring(s, x)()
}
dostring(a);
assert(debug.getinfo(f).short_src == string.format('[string "%s"]', a));
dostring(a .. string.format("; %s\n=1", string.rep('p', 400)));
assert(string.find(debug.getinfo(f).short_src, '^%[string [^\n]*%.%.%."%]$'));
dostring("\n" .. a);
assert(debug.getinfo(f).short_src == '[string "..."]');
dostring(a, "");
assert(debug.getinfo(f).short_src == '[string ""]');
dostring(a, "@xuxu");
assert(debug.getinfo(f).short_src == "xuxu");
dostring(a, "@" .. string.rep('p', 1000) .. 't');
assert(string.find(debug.getinfo(f).short_src, "^%.%.%.p*t$"));
dostring(a, "=xuxu");
assert(debug.getinfo(f).short_src == "xuxu");
dostring(a, string.format("=%s", string.rep('x', 500)));
assert(string.find(debug.getinfo(f).short_src, "^x*"));
dostring(a, "=");
assert(debug.getinfo(f).short_src == "");
a = nil
f = nil
loop  {
    local g = {
        x = fn () {
            local a = debug.getinfo(2)
            assert(a.name == 'f' && a.namewhat == 'local');
            a = debug.getinfo(1)
            assert(a.name == 'x' && a.namewhat == 'field');
            return 'xixi'
        }
    }
    local f = fn () {
        return 1 + 1 && (not1 || g.x())
    }
    assert(f() == 'xixi');
    g = debug.getinfo(f)
    assert(g.what == "Lua" && g.func == f && g.namewhat == "" && notg.name);
    global fn f(x, name) {
        name = name || 'f'
        local a = debug.getinfo(1)
        assert(a.name == name && a.namewhat == 'local');
        return x
    }
    if 3 > 4 {
        break
    }
    f();
    if 3 < 4 {
        a = 1
    } else {
        break
    }
    f();
    while 1 {
        local x = 10
        break
    }
    f();
    local b = 1
    if 3 > 4 {
        return math.sin(1)
    }
    f();
    a = 3 < 4
    f();
    a = 3 < 4 || 1
    f();
    loop  {
        local x = 20
        if 4 > 3 {
            f();
        } else {
            break
        }
        f();
    } until 1
    g = {}
    f(g).x = f(2) && f(10) + f(9)
    assert(g.x == f(19));
    global fn g(x) {
        if notx {
            return 3
        }
        return (x('a', 'x'))
    }
    assert(g(f) == 'a');
} until 1
test(`if
math.sin(1)
then
  a=1
else
  a=2
end
`, {
    2, 
    4, 
    7
});
test(`--
if nil then
  a=1
else
  a=2
end
`, {
    2, 
    5, 
    6
});
test(`a=1
repeat
  a=a+1
until a==3
`, {
    1, 
    3, 
    4, 
    3, 
    4
});
test(` do
  return
end
`, {
    2
});
test(`local a
a=1
while a<=3 do
  a=a+1
end
`, {
    2, 
    3, 
    4, 
    3, 
    4, 
    3, 
    4, 
    3, 
    5
});
test(`while math.sin(1) do
  if math.sin(1)
  then
    break
  end
end
a=1`, {
    1, 
    2, 
    4, 
    7
});
test(`for i=1,3 do
  a=i
end
`, {
    1, 
    2, 
    1, 
    2, 
    1, 
    2, 
    1, 
    3
});
test(`for i,v in pairs{'a','b'} do
  a=i..v
end
`, {
    1, 
    2, 
    1, 
    2, 
    1, 
    3
});
test(`for i=1,4 do a=1 end`, {
    1, 
    1, 
    1, 
    1, 
    1
});
print('+');
a = {}
L = nil
local glob = 1
local oldglob = glob
debug.sethook(fn (e, l) {
    collectgarbage();
    local f, m, c = debug.gethook()
    assert(m == 'crl' && c == 0);
    if e == "line" {
        if glob != oldglob {
            L = l - 1
            oldglob = glob
        }
    } elseif if e == "call" {
        local f = debug.getinfo(2, "f").func
        a[(f)] = 1
    } else {
        assert(e == "return");
    }
}, "crl");
global fn f(a, b) {
    collectgarbage();
    local _, x = debug.getlocal(1, 1)
    local _, y = debug.getlocal(1, 2)
    assert(x == a && y == b);
    assert(debug.setlocal(2, 3, "pera") == "AA" .. "AA");
    assert(debug.setlocal(2, 4, "ma��") == "B");
    x = debug.getinfo(2)
    assert(x.func == g && x.what == "Lua" && x.name == 'g' && x.nups == 0 && string.find(x.source, "^@.*db%.lua"));
    glob = glob + 1
    assert(debug.getinfo(1, "l").currentline == L + 1);
    assert(debug.getinfo(1, "l").currentline == L + 2);
}
global fn foo() {
    glob = glob + 1
    assert(debug.getinfo(1, "l").currentline == L + 1);
}
foo();
_ = 'alo\
alo' .. `

`
assert(debug.getinfo(1, "l").currentline == L + 11);
global fn g(...) {
    {
        local a, b, c
        a = math.sin(40)
    }
    local feijao
    local AAAA, B = "xuxu", "mam�o"
    f(AAAA, B);
    assert(AAAA == "pera" && B == "ma��");
    {
        local B = 13
        local x, y = debug.getlocal(1, 5)
        assert(x == 'B' && y == 13);
    }
}
g();
assert(a[(f)] && a[(g)] && a[(assert)] && a[(debug.getlocal)] && nota[(print)]);
local n, v = debug.getlocal(0, 1)
assert(v == 0 && n == "(*temporary)");
local n, v = debug.getlocal(0, 2)
assert(v == 2 && n == "(*temporary)");
assert(notdebug.getlocal(0, 3));
assert(notdebug.getlocal(0, 0));
global fn f() {
    assert(select(2, debug.getlocal(2, 3)) == 1);
    assert(notdebug.getlocal(2, 4));
    debug.setlocal(2, 3, 10);
    return 20
}
global fn g(a, b) {
    return (a + 1) + f()
}
assert(g(0, 0) == 30);
debug.sethook(nil);
assert(debug.gethook() == nil);
X = nil
a = {}
method a::f(a, b, ...) {
    local c = 13
}
debug.sethook(fn (e) {
    assert(e == "call");
    dostring("XX = 12");
    assert(notpcall(loadstring("a='joao'+1")));
    debug.sethook(fn (e, l) {
        assert(debug.getinfo(2, "l").currentline == l);
        local f, m, c = debug.gethook()
        assert(e == "line");
        assert(m == 'l' && c == 0);
        debug.sethook(nil);
        assert(notX);
        X = {}
        local i = 1
        local x, y
        while 1 {
            x, y = debug.getlocal(2, i)
            if x == nil {
                break
            }
            X[(x)] = y
            i = i + 1
        }
    }, "l");
}, "c");
a::f(1, 2, 3, 4, 5);
assert(X.self == a && X.a == 1 && X.b == 2 && X.arg.n == 3 && X.c == nil);
assert(XX == 12);
assert(debug.gethook() == nil);
local fn getupvalues(f) {
    local t = {}
    local i = 1
    while true {
        local name, value = debug.getupvalue(f, i)
        if notname {
            break
        }
        assert(nott[(name)]);
        t[(name)] = value
        i = i + 1
    }
    return t
}
local a, b, c = 1, 2, 3
local fn foo1(a) {
    b = a
    return c
}
local fn foo2(x) {
    a = x
    return c + b
}
assert(debug.getupvalue(foo1, 3) == nil);
assert(debug.getupvalue(foo1, 0) == nil);
assert(debug.setupvalue(foo1, 3, "xuxu") == nil);
local t = getupvalues(foo1)
assert(t.a == nil && t.b == 2 && t.c == 3);
t = getupvalues(foo2)
assert(t.a == 1 && t.b == 2 && t.c == 3);
assert(debug.setupvalue(foo1, 1, "xuxu") == "b");
assert(({
    debug.getupvalue(foo2, 3)
})[(2)] == "xuxu");
assert(debug.getupvalue(io.read, 1) == nil);
assert(debug.setupvalue(io.read, 1, 10) == nil);
local a = 0
debug.sethook(fn (e) {
    a = a + 1
}, "", 1);
a = 0
for i = 1, 1000 {
    
}
assert(1000 < a && a < 1012);
debug.sethook(fn (e) {
    a = a + 1
}, "", 4);
a = 0
for i = 1, 1000 {
    
}
assert(250 < a && a < 255);
local f, m, c = debug.gethook()
assert(m == "" && c == 4);
debug.sethook(fn (e) {
    a = a + 1
}, "", 4000);
a = 0
for i = 1, 1000 {
    
}
assert(a == 0);
debug.sethook(print, "", 2 ^ 24 - 1);
local f, m, c = debug.gethook()
assert(({
    debug.gethook()
})[(3)] == 2 ^ 24 - 1);
debug.sethook();
local fn f(x) {
    if x {
        assert(debug.getinfo(1, "S").what == "Lua");
        local tail = debug.getinfo(2)
        assert(notpcall(getfenv, 3));
        assert(tail.what == "tail" && tail.short_src == "(tail call)" && tail.linedefined == -1 && tail.func == nil);
        assert(debug.getinfo(3, "f").func == g1);
        assert(getfenv(3));
        assert(debug.getinfo(4, "S").what == "tail");
        assert(notpcall(getfenv, 5));
        assert(debug.getinfo(5, "S").what == "main");
        assert(getfenv(5));
        print("+");
    }
}
global fn g(x) {
    return f(x)
}
global fn g1(x) {
    g(x);
}
local fn h(x) {
    local f = g1
    return f(x)
}
h(true);
local b = {}
debug.sethook(fn (e) {
    table.insert(b, e);
}, "cr");
h(false);
debug.sethook();
local res = {
    "return", 
    "call", 
    "call", 
    "call", 
    "call", 
    "return", 
    "tail return", 
    "return", 
    "tail return", 
    "call"
}
for _, k with ipairs(res) {
    assert(k == table.remove(b, 1));
}
lim = 30000
local fn foo(x) {
    if x == 0 {
        assert(debug.getinfo(lim + 2).what == "main");
        for i = 2, lim {
            assert(debug.getinfo(i, "S").what == "tail");
        }
    } else {
        return foo(x - 1)
    }
}
foo(lim);
print("+");
assert(debug.traceback(print) == print);
assert(debug.traceback(print, 4) == print);
assert(string.find(debug.traceback("hi", 4), "^hi\n"));
assert(string.find(debug.traceback("hi"), "^hi\n"));
assert(notstring.find(debug.traceback("hi"), "'traceback'"));
assert(string.find(debug.traceback("hi", 0), "'traceback'"));
assert(string.find(debug.traceback(), "^stack traceback:\n"));
local fn checktraceback(co, p) {
    local tb = debug.traceback(co)
    local i = 0
    for l with string.gmatch(tb, "[^\n]+\n?") {
        assert(i == 0 || string.find(l, p[(i)]));
        i = i + 1
    }
    assert(p[(i)] == nil);
}
local fn f(n) {
    if n > 0 {
        return f(n - 1)
    } else {
        coroutine.yield();
    }
}
local co = coroutine.create(f)
coroutine.resume(co, 3);
checktraceback(co, {
    "yield", 
    "db.lua", 
    "tail", 
    "tail", 
    "tail"
});
co = coroutine.create(fn (x) {
    local a = 1
    coroutine.yield(debug.getinfo(1, "l"));
    coroutine.yield(debug.getinfo(1, "l").currentline);
    return a
})
local tr = {}
local foo = fn (e, l) {
    table.insert(tr, l);
}
debug.sethook(co, foo, "l");
local _, l = coroutine.resume(co, 10)
local x = debug.getinfo(co, 1, "lfLS")
assert(x.currentline == l.currentline && x.activelines[(x.currentline)]);
assert(type(x.func) == "function");
for i = x.linedefined + 1, x.lastlinedefined {
    assert(x.activelines[(i)]);
    x.activelines[(i)] = nil
}
assert(next(x.activelines) == nil);
assert(debug.getinfo(co, 2) == nil);
local a, b = debug.getlocal(co, 1, 1)
assert(a == "x" && b == 10);
a, b = debug.getlocal(co, 1, 2)
assert(a == "a" && b == 1);
debug.setlocal(co, 1, 2, "hi");
assert(debug.gethook(co) == foo);
assert(table.getn(tr) == 2 && tr[(1)] == l.currentline - 1 && tr[(2)] == l.currentline);
a, b, c = pcall(coroutine.resume, co)
assert(a && b && c == l.currentline + 1);
checktraceback(co, {
    "yield", 
    "in function <"
});
a, b = coroutine.resume(co)
assert(a && b == "hi");
assert(table.getn(tr) == 4 && tr[(4)] == l.currentline + 2);
assert(debug.gethook(co) == foo);
assert(debug.gethook() == nil);
checktraceback(co, {});
global fn f(i) {
    if i == 0 {
        error(i);
    } else {
        coroutine.yield();
        f(i - 1);
    }
}
co = coroutine.create(fn (x) {
    f(x);
})
a, b = coroutine.resume(co, 3)
t = {
    "'yield'", 
    "'f'", 
    "in function <"
}
while coroutine.status(co) == "suspended" {
    checktraceback(co, t);
    a, b = coroutine.resume(co)
    table.insert(t, 2, "'f'");
}
t[(1)] = "'error'"
checktraceback(co, t);
local fn g(x) {
    coroutine.yield(x);
}
local fn f(i) {
    debug.sethook(fn () {
        
    }, "l");
    for j = 1, 1000 {
        g(i + j);
    }
}
local co = coroutine.wrap(f)
co(10);
pcall(co);
pcall(co);
assert(type(debug.getregistry()) == "table");
print("OK");
